### PR TITLE
DAOS-6854 dfuse: Use dfs_ostatx for aync getattr calls.

### PR DESCRIPTION
**Pull request by ashleypittman**
Skip-build: true

Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/client/dfuse/ops/fgetattr.c [hunk 1]
```diff
@@ -1,33 +1,64 @@
...
+		D_GOTO(err, rc = ENOMEM);
 
-	attr.st_ino = ie->ie_stat.st_ino;
+	ev->de_req         = req;
+	ev->de_complete_cb = dfuse_cb_getattr_cb;
+	ev->de_ie          = ie;
 
-	ie->ie_stat = attr;
+	rc = daos_event_init(&ev->de_ev, eqt->de_eq, NULL);
+	if (rc != -DER_SUCCESS)
```
Comment (mchaarawi) [inline src/client/dfuse/ops/fgetattr.c:54]: need to free ev here?

File: src/client/dfuse/ops/fgetattr.c [hunk 2]
```diff
@@ -1,33 +1,64 @@
...
+		D_GOTO(err, rc = daos_der2errno(rc));
+
+	rc = dfs_ostatx(ie->ie_dfs->dfs_ns, ie->ie_obj, &ev->de_attr, &ev->de_ev);
+	if (rc != 0)
```
Comment (mchaarawi) [inline src/client/dfuse/ops/fgetattr.c:58]: same

**Review by mchaarawi — COMMENTED** [review]
is there any indication of performance gain with this?

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -4939,6 +4939,197 @@ dfs_ostat(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf)
...
+	d_iov_set(&ess->sg_iovs[i++], &ess->entry.uid, sizeof(uid_t));
+	d_iov_set(&ess->sg_iovs[i++], &ess->entry.gid, sizeof(gid_t));
+	d_iov_set(&ess->sg_iovs[i++], &ess->entry.value_len, sizeof(daos_size_t));
+	d_iov_set(&ess->sg_iovs[i++], &ess->entry.obj_hlc, sizeof(uint64_t));
+
+	ess->sgl.sg_nr     = i;
+	ess->sgl.sg_nr_out = 0;
+	ess->sgl.sg_iovs   = ess->sg_iovs;
+
+	rc = daos_event_register_comp_cb(ev, dfs_ostat_cb, ess);
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: this was a draft comment from an earlier review. disregard

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -4939,6 +4939,197 @@ dfs_ostat(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf)
...
+#if 0
+	if (ess->cb_count++ == 1)
+		dfs_entry_cb(ess, ev, rc);
+#endif
+
+	return 0;
+}
+
+int
+dfs_ostatx(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, daos_event_t *ev)
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: this was a draft comment from an earlier review. disregard

**Comment by ashleypittman** [comment]
> is there any indication of performance gain with this?

It's complicated.  As much as we can take numbers from our CI systems some tests are much faster, some are a little slower.

The latest master build:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/master/8998/testReport/FTEST_dfuse/DaosBuild/
This PR:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-12832/14/testReport/FTEST_dfuse/DaosBuild/

The `nocache` version which I'd expect to benefit most from this went down from 1h,19m to 41m so almost a 50% speedup, the `wb` cache test however went from 7m,27s to 8m,8s so perhaps 10% slower.

I'd expect the nocache test to get much faster, this was the test that triggered this work but 50% is hard to explain and other than the fact that the event queue is shared with reads/writes I don't have an explanation for why the wb test is slower.

File: src/client/dfuse/ops/fgetattr.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/fgetattr.c:54]: Yes, will re-push.

Comment (mchaarawi) [reply src/client/dfuse/ops/fgetattr.c:54]: you said you will repush to fix this, but sounds like still needs the fix.

**Review by mchaarawi — CHANGES_REQUESTED** [review]
just need to free the ev on failure case

File: src/client/dfuse/ops/fgetattr.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/fgetattr.c:54]: Ah yes, sorry.  I've got too many PRs in flight at the mintue.

**Review by mchaarawi — APPROVED** [review]

**Review by wiliamhuang — APPROVED** [review]

**Pull request merged**